// @icp-sdk/core/lib/cjs/identity/identity/attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttributesIdentity = void 0;
/**
 * An Identity decorator that injects `sender_info` into the request body
 * before delegating to an inner identity for signing.
 *
 * Because `sender_info` is part of the request content, it is included in the
 * representation-independent hash (`requestIdOf`) and covered by the sender's
 * signature.
 *
 * @example
 * ```ts
 * const inner = DelegationIdentity.fromDelegation(key, chain);
 * const identity = new AttributesIdentity({
 *   inner,
 *   attributes: { data: new Uint8Array([...]), signature: new Uint8Array([...]) },
 *   signer: { canisterId: Principal.fromText('aaaaa-aa') },
 * });
 * const agent = HttpAgent.create({ identity });
 * ```
 */
class AttributesIdentity {
    #inner;
    #attributes;
    #signer;
    /**
     * @param options - Configuration for the identity.
     * @param options.inner - The inner identity to delegate signing to.
     * @param options.attributes - The signed attributes to include in the request.
     * @param options.signer - The canister that signed the attributes.
     */
    constructor(options) {
        this.#inner = options.inner;
        this.#attributes = options.attributes;
        this.#signer = options.signer;
    }
    getPrincipal() {
        return this.#inner.getPrincipal();
    }
    transformRequest(request) {
        return this.#inner.transformRequest({
            ...request,
            body: {
                ...request.body,
                sender_info: {
                    signer: this.#signer.canisterId.toUint8Array(),
                    info: this.#attributes.data,
                    sig: this.#attributes.signature,
                },
            },
        });
    }
}
exports.AttributesIdentity = AttributesIdentity;
